ui/app/services/actors-relationships.js

```diff
@@ -0,0 +1,95 @@
+import Service from '@ember/service';
+import { action } from '@ember/object';
+import { tracked } from '@glimmer/tracking';
+import { schedule } from '@ember/runloop';
+import { getBoxToBoxArrow } from 'curved-arrows';
+
+function boxToArrow(ra, rb) {
+  const bbA = ra;
+  const bbB = rb;
+
+  const [sx, sy, c1x, c1y, c2x, c2y, ex, ey, ae, as] = getBoxToBoxArrow(
+    bbA.offsetLeft,
+    bbA.offsetTop,
+    bbA.offsetWidth,
+    bbA.offsetHeight,
+    bbB.offsetLeft,
+    bbB.offsetTop,
+    bbA.offsetWidth,
+    bbB.offsetHeight
+  );
+
+  return {
+    sx,
+    sy,
+    c1x,
+    c1y,
+    c2x,
+    c2y,
+    ex,
+    ey,
+    ae,
+    as,
+  };
+}
+
+export default class ActorRelationshipService extends Service {
+  @tracked actors = [];
+
+  get fns() {
+    const { registerActor, deregisterActor, recalcCurves } = this;
+    return { registerActor, deregisterActor, recalcCurves };
+  }
+
+  get data() {
+    const { actors, relationships } = this;
+    return { actors, relationships };
+  }
+
+  @action registerActor(actor) {
+    schedule('actions', this, () => {
+      this.actors = [...this.actors, actor];
+    });
+  }
+
+  @action deregisterActor(actor) {
+    schedule('actions', this, () => {
+      this.actors = this.actors.filter((a) => a !== actor);
+    });
+  }
+
+  get rects() {
+    const { actors } = this;
+
+    return actors
+      .filter((e) => e.previousEval) // should we make this more generic
+      .map((e) => {
+        const { previousEval: pid, id } = e;
+
+        const eRectangle = document.querySelector(`[data-eval="${id}"]`);
+        const prevRectangle = document.querySelector(`[data-eval="${pid}"]`);
+
+        return [eRectangle, prevRectangle];
+      });
+    // find previousEval if it exists calculate and add add a path
+  }
+
+  get relationships() {
```

Comment:
When we save the start and endpoint we can draw circles to make the start/endpoints appear like ports:

<img width="591" alt="ports" src="https://user-images.githubusercontent.com/242299/159955232-144a7422-399f-4e08-a6e2-2d2c9a9e670e.png">

```js
  get relationships() {

    const { rects } = this;

    return rects.map(([eRectangle, prevRectangle]) => {
      const { sx, sy, c1x, c1y, c2x, c2y, ex, ey } = boxToArrow(
        eRectangle,
        prevRectangle
      );

      return {
        d: `M ${sx} ${sy} C ${c1x} ${c1y}, ${c2x} ${c2y}, ${ex} ${ey}`,
        sx,
        sy,
        ex,
        ey,
      };
    });
  }

```

And then in the svg markup:

```hbs
              <Providers::ActorsRelationships as |actors|>
                <svg
                  width={{this.width}}
                  height={{this.height}}
                  style="z-index: 10; inset: 0; position: absolute; pointer-events: none"
                  {{did-update actors.fns.recalcCurves this.width}}
                >
                  {{#each actors.data.relationships as |r|}}
                    <path
                      d={{r.d}}
                      stroke="#7E8FA8"
                      strokeWidth="1"
                      fill="none"
                    ></path>
                    <circle
                      cx={{r.sx}}
                      cy={{r.sy}}
                      r="4"
                      fill="white"
                      stroke="black"
                    ></circle>
                    <circle
                      cx={{r.ex}}
                      cy={{r.ey}}
                      r="4"
                      fill="white"
                      stroke="black"
                    ></circle>
                  {{/each}}
                </svg>
              </Providers::ActorsRelationships>
```